src/main/java/org/apache/sling/jcr/repoinit/impl/RepositoryInitializerFactory.java

```diff
@@ -122,14 +131,47 @@ public void processRepository(final SlingRepository repo) throws Exception {
                             continue;
                         }
                         final List<Operation> ops = parser.parse(new StringReader(script));
-                        log.info("Executing {} repoinit operations", ops.size());
-                        processor.apply(s, ops);
-                        s.save();
+                        String msg = String.format("Executing %s repoinit operations", ops.size());
+                        log.info(msg);
+                        applyOperations(s,ops,msg);
                     }
                 }
             } finally {
                 s.logout();
             }
         }
     }
+
+
+    /**
+     * Apply the operations within a session, support retries
+     * @param session the JCR session to use
+     * @param ops the list of operations
+     * @param logMessage the messages to print when retry
+     * @throws Exception if the application fails despite the retry
+     */
+    private void applyOperations(Session session, List<Operation> ops, String logMessage) throws Exception {
+
+        RetryableOperation retry = new RetryableOperation.Builder().withBackoffBase(1000).withMaxRetries(3).build();
+        boolean successful = retry.apply(() -> {
+            try {
+                processor.apply(session, ops);
+                session.save();
+                return true;
+            } catch (RepositoryException e) {
+                log.error("(temporarily) failed to apply repoinit operations",e);
```

Comment:
So in case of permanent failure we end up with the same exception being logged N (=3) times. We should avoid that

Reply:
Yes, that is possible, but I don't think it makes sense to suppress it, because we are never sure that it will always the same reason and the same exception. It's completely ok if the different RepositoryExceptions are thrown.

Anyway, I don't expect this situation to happen that often, as we have seen it only in rare circumstances. So that's for me an acceptable logging overhead.

Reply:
is now adjusted to log only the last exception.